core/src/main/java/com/radixdlt/genesis/olympia/OlympiaEndStateApiClient.java

```diff
@@ -0,0 +1,166 @@
+  private static final String STATUS_READY = "ready";
+  private static final String STATUS_NOT_READY = "not_ready";
+
+  public sealed interface OlympiaEndStateResponse {
+    record Ready(String signature, String hash, String base64Contents)
+        implements OlympiaEndStateResponse {}
+
+    record NotReady(String signature, String placeholderHash) implements OlympiaEndStateResponse {}
+  }
+
+  private final URL endStateApiUrl;
+  private final Network network;
+
+  @Inject
+  public OlympiaEndStateApiClient(OlympiaGenesisConfig config, Network network) {
+    try {
+      this.endStateApiUrl = new URL(config.nodeCoreApiUrl(), END_STATE_API_PATH);
+    } catch (MalformedURLException e) {
+      throw new RuntimeException(
+          // Shouldn't really happen, nodeCoreApiUrl is verified earlier
+          String.format(
+              """
+                  Failed to create OlympiaEndStateApiClient. \
+                  Invalid endStateApiUrl (coreApiUrl=%s path=%s)""",
+              config.nodeCoreApiUrl(), END_STATE_API_PATH));
+    }
+    this.network = network;
+  }
+
+  public OlympiaEndStateResponse getOlympiaEndState() throws IOException, JSONException {
+    final var requestData =
+        new JSONObject()
+            .put(
+                "network_identifier",
+                new JSONObject().put("network", toOlympiaNetworkIdentifier(network)));
+
+    final var response = postForJson(endStateApiUrl, requestData);
+
+    final var status = response.getString(STATUS_FIELD);
+
+    switch (status) {
+      case STATUS_READY -> {
+        return new OlympiaEndStateResponse.Ready(
+            response.getString("signature"),
+            response.getString("hash"),
+            response.getString("contents"));
+      }
+      case STATUS_NOT_READY -> {
+        return new OlympiaEndStateResponse.NotReady(
+            response.getString("signature"), response.getString("placeholder_hash"));
+      }
+      default -> throw new RuntimeException("Unexpected response status " + status);
+    }
+  }
+
+  private static JSONObject postForJson(URL url, JSONObject request) throws IOException {
+    final var conn = (HttpURLConnection) url.openConnection();
+    conn.setRequestMethod("POST");
+    conn.setRequestProperty("Content-Type", "application/json");
+    conn.setDoOutput(true);
+    final var requestPayload = request.toString();
+    try (var os = conn.getOutputStream()) {
+      byte[] input = requestPayload.getBytes(StandardCharsets.UTF_8);
+      os.write(input, 0, input.length);
+    }
+    final var responseCode = conn.getResponseCode();
+    if (responseCode != 200) {
+      throw new RuntimeException("Unexpected response code: " + responseCode);
+    }
+    try (var br =
+        new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))) {
+      final var response = new StringBuilder();
+      String responseLine;
+      while ((responseLine = br.readLine()) != null) {
```

Comment:
Hmmm there probably aren't separate lines here because most of this is a massive hex string...

This is probably fine, but should probably be tested?

Reply:
I'd feel more comfortable buffering into a JSON parser.. or even better, just using a library like Jakub suggests.

We already use a library for the integration tests 👍